### ISSUE TITLE
Significant degradation between versions

### ISSUE THREAD

**Issue by keveman** [HomebrewML/HeavyBall#58]
<img width="917" alt="Image" src="https://github.com/user-attachments/assets/5e449c89-0264-4a55-bffc-d3996cef1e9d" />
I am attaching a graph comparing loss curves for my model across two commits of HeavyBall be36047 and c48bcaf
There was no other change in my code, not in the model or data. My call to the optimizer looks as follows:

```
optimizer = heavyball.ForeachPSGDKron(
        opt_grouped_params,
        lr=2e-3,
        weight_decay=5e-4,
        warmup_steps=1024*16,
)
```

Granted c48bcaf is not current HEAD, but I was taken aback by this sudden and drastic loss of performance. @ClashLuke Any hints?

**Comment by keveman** [comment]
@ClashLuke This is mostly a fact finding exercise for me and I don't want you to treat this as a bug. I'm yet to try HEAD on my model, but waiting for some GPU resources.

**Comment by ClashLuke** [comment]
Could you try setting `precond_init_scale=1`?

**Comment by keveman** [comment]
Thanks for looking into this @ClashLuke. I am not sure whether you meant trying `precond_init_scale=1` at c48bcaf or HEAD, but I'll try both and report back as soon as I get some GPU time.

**Comment by ClashLuke** [comment]
HEAD and c48bcaf don't a have any significant differences for PSGD, so, it's definitely an unfixed bug. Hopefully we can fix it soon

**Comment by keveman** [comment]
Sorry it took this long, but here is the comparison with `precond_init_scale=1`.
<img width="855" alt="Image" src="https://github.com/user-attachments/assets/a6fc1d6d-e18e-4e3a-9433-5f254b3df909" />
No change pretty much with `precond_init_scale=1`, so definitely something else changed after be36047 causing the dramatic difference.

**Comment by ClashLuke** [comment]
Could you share any information about the problem to help me reproduce the curves? (or modify one of the example files to match your setup)

Otherwise, the other major changes can be disabled using
1) `dampening=0`
2) `lower_bound_beta=0`
3) `precond_update_power_iterations=1`

The updated PSGD also balances the preconditioners at every step, slightly changing learning dynamics. This didn't break anything in my testing, but you could comment the `psgd_balance_Q` lines here:
https://github.com/HomebrewML/HeavyBall/blob/c48bcafebbc11ea77db0c6a06387cb7a45987785/heavyball/utils.py#L2025-L2031

**Comment by keveman** [comment]
Thanks for looking into this @ClashLuke. The model I running is here https://huggingface.co/UsefulSensors/moonshine-tiny
I am running on 64 GPUs with the `accelerate` framework. I'll try to repro this with single GPU setting, and also try modify the example files like you suggested, and report back.

**Comment by ClashLuke** [comment]
Can't reproduce a regression with [autoencoder.py](https://github.com/HomebrewML/HeavyBall/blob/5ddfea278accb9b5a8013651f32e1c23625b0b9c/examples/autoencoder.py) across various model and batch sizes, as well as with and without the flags above and independently testing `cache` (the only other thing that was changed) - the curves are almost identical.

I did notice that `stochastic_schedule` was changed from `True` to `False` and that the corresponding argument was ignored, which is now fixed

The only noteworthy things seem to be the flags mentioned above

**Comment by ClashLuke** [comment]
Sorry, could you clarify your message above @anhnami ? Does heavyball.ForeachAdamW converge slower than torch.optim.AdamW? If so, please make sure the hyperparameters match.

This is not documented, but heavyball has different defaults than torch.optim, optax and others. heavyball's implementation is aligned closely with that of torch, except for slight precision improvements. However, its beta2 is set to 0.99 compared to torch's 0.999.

**Comment by anhnami** [comment]
Hi, sorry for the late reply. In May, we trained Conformer RNN-T ASR model with heavyball.ForeachAdamW and torch.optim.AdamW using identical settings: lr=1e-3, wd=1e-3, betas[0.9, 0.98], batch size: 2048. For some unknown reasons, ForeachAdamW converged much slower than AdamW (similar to training curves showed by keveman). Unfortunately, I have removed the training logs for them.

**Comment by ClashLuke** [comment]
PSGD's implementation is now closely aligned with [the new, official implementation](https://github.com/lixilinx/psgd_torch/blob/89b4cead31b7ad1494c4cf4dc39f4cbf920ff14d/psgd.py). The math matches, with the most significant difference being that we do more power iterations to get a more precise estimate of the largest singular value.

I'm closing this now, but feel free to reopen it if the errors persist.
